### PR TITLE
2.0 IDEX Independent z offset and other fixes

### PR DESCRIPTION
**Pull request by InsanityAutomation**
See discussion in #11412 - misc idex fixes and independent z offset for each extruder

### CODE REVIEW

**Comment by Roxy-3D** [comment]
I think we should change the ICSP #if stuff.   You edited the pin numbers.   But the thing is....  The ICSP port pins seem to be 'standard' on 50, 51, 52.    It probably would be better to just add your own special case #define to select a different, one off set of pins.

But...    We can certainly discuss and debate that!

**Comment by InsanityAutomation** [comment]
@Roxy-3D What i put in is for the branch with icsp enabled, and follows an accessible plug for a single sensor. Its what a generic user could do without major hacking that you should expect them to know what pin theyre hacking on. Could be debated either way for what to set it to, but neither ships from formbot with that use, so ultimately correct is probably removing it altogether. Most machines Chris sells nowadays do have the runout sensor on the pin defined here though, so as the only authorized distributor with that function bundled and therefore a large number of machines configured that way, my first thought was to keep it in the stock config.

**Comment by Roxy-3D** [comment]
I understand.   My point is the ICSP port (which is available on the FORMBOT v1 boards) uses pins 50, 51 and 52,  I have not searched through a huge number of boards but the Atmega-2560 uses the same pin numbers for its ICSP port.   And so does the UNO.

My thinking is...   That port IS available to Formbot users.   They have to give up the use of the SD Memory card to use it.  But probably we shouldn't have incorrect pin numbers within the ICSP #ifdef.   Instead, I was suggesting we may want to have an extra #ifdef that handles the case you are trying to provide pins to handle.


More comments on other things as time permits!

**Comment by Roxy-3D** [comment]
Is this correct?
```
    #if ENABLED(INDEPENDENT_Z_OFFSETS) && ENABLED(DUAL_X_CARRIAGE)
      _FIELD_TEST(x1_zprobe_zoffset);
      _FIELD_TEST(x2_zprobe_zoffset);
      EEPROM_WRITE(x1_zprobe_zoffset);
      EEPROM_WRITE(x1_zprobe_zoffset);
      EEPROM_WRITE(x2_zprobe_zoffset);
    #else
```
Is the first write of x1_zprobe_zoffset needed?   Should we be removing the zprobe_zoffset if we are in this type of setup?

**Comment by InsanityAutomation** [comment]
@Roxy-3D Since on startup T0 is used, and at save T1 maybe active, we need to make sure that we align it to the active tool head. If we used the zprobe zoffset when t1 was active, the wrong number would be written for the tool active on startup.

**Comment by Roxy-3D** [comment]
Yes.  I understood that part.   But if we are in Dual X Carraige mode, and have independent Z-Offsets enabled....   Wouldn't it make sense to always use T0's or T1's number and not even have the zprobe_zoffset number?   Or maybe we keep zprobe_zoffset active in the code, but populate its value from T0's number in this mode? (But don't store it in EEPROM)

Oh... and one other question...   That Z-Height problem with tool changes back and forth...   Is that fixed in the current code base?   Or is that something you are still looking at?

**Comment by InsanityAutomation** [comment]
We could eliminate storing x1_zprobe_zoffset and use the base value for x1 since at save and restore they should be identical. At runtime zprobe_zoffset switches with active tool as it drives the physical movement so mainly I wanted to avoid confusion and keep them separate.

**Comment by InsanityAutomation** [comment]
@Roxy-3D LMK if the single value is the way you want to go. If so, Ill update that when I get a few minute break here.

**Comment by Roxy-3D** [comment]
I'm not hung up on it being either way.   I was just asking so I could understand better.    This is just an opinion....  But I think the 'right thing' to do is not store the zprobe_zoffset value in the Dual X Carriage case with independent Z Offsets.    And then in the conditional code for that case, just read T0's value and T1's value.   AND then use T0's value to populate zprobe_zoffset to the same value as T0.

But...   I don't have strong feelings on that.

**Comment by InsanityAutomation** [comment]
@Roxy-3D changed for you

**Comment by Roxy-3D** [comment]
I looked through everything...   How about we remove the ICSP pin defines?  That would be better than having them be wrong.   If I use that port to get a few signals...   I can just turn off the SD Memory Card and then plug into them.

**Comment by InsanityAutomation** [comment]
Sounds good. I'll make that revision and add it here shortly, then we can get this merged.

**Comment by InsanityAutomation** [comment]
@Roxy-3D as soon as Travis is done, this should be good to go.

**Comment by Roxy-3D** [comment]
I think we should let @thinkyhead merge this.   The reason is he might want to weigh in on the EEPROM storage.   But my guess is he will merge it quickly.

**Comment by thinkyhead** [comment]
> Is this correct?

No. The `_FIELD_TEST` goes immediately in front of the field about to be read or written. You can't put two of them in a row as shown.

File: Marlin/Configuration_adv.h [hunk 1]
```diff
@@ -372,7 +372,9 @@
       // override for X2_HOME_POS. This also allow recalibration of the distance between the two endstops
       // without modifying the firmware (through the "M218 T1 X???" command).
       // Remember: you should set the second extruder x-offset to 0 in your slicer.
-
+  //#define INDEPENDENT_Z_OFFSETS   // Allows settins a seperate z offset for X2. This allows compensation
+      //for tool heads that are at different heights. Please note this will be disabled in duplicate mode
+      // and the X1 z offset will be used
```
Comment (thinkyhead) [inline Marlin/Configuration_adv.h]: Spelling and formatting count, and this needs a bit of attention.

File: Marlin/src/gcode/feature/pause/M600.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ void GcodeSuite::M600() {
     int8_t DXC_ext = target_extruder;
     if (!parser.seen('T')) {  // If no tool index is specified, M600 was (probably) sent in response to filament runout.
                               // In this case, for duplicating modes set DXC_ext to the extruder that ran out.
-      #if ENABLED(FILAMENT_RUNOUT_SENSOR)
+      #if ENABLED(FILAMENT_RUNOUT_SENSOR) && NUM_RUNOUT_SENSORS>1
```
Comment (thinkyhead) [inline Marlin/src/gcode/feature/pause/M600.cpp]: There should be spaces around operators.

Please read http://marlinfw.org/docs/development/coding_standards.html asap.

File: Marlin/src/gcode/motion/M290.cpp [hunk 3]
```diff
@@ -37,6 +37,13 @@
 #if ENABLED(BABYSTEP_ZPROBE_OFFSET)
   FORCE_INLINE void mod_zprobe_zoffset(const float &offs) {
     zprobe_zoffset += offs;
+    #if ENABLED(DUAL_X_CARRIAGE) && ENABLED(INDEPENDENT_Z_OFFSETS)
+      #if(active_extruder==0)
+        x1_zprobe_zoffset = zprobe_zoffset;
+      #else
+        x2_zprobe_zoffset = zprobe_zoffset;
+      #endif
+    #endif
```
Comment (thinkyhead) [inline Marlin/src/gcode/motion/M290.cpp]: This is not valid C++ code.

File: Marlin/src/gcode/probe/M851.cpp [hunk 4]
```diff
@@ -32,7 +32,16 @@ void GcodeSuite::M851() {
...
     if (WITHIN(value, Z_PROBE_OFFSET_RANGE_MIN, Z_PROBE_OFFSET_RANGE_MAX))
+    {
       zprobe_zoffset = value;
+      #if ENABLED(DUAL_X_CARRIAGE) && ENABLED(INDEPENDENT_Z_OFFSETS)
+        #if(active_extruder==0)
+          x1_zprobe_zoffset = zprobe_zoffset;
+        #else
+          x2_zprobe_zoffset = zprobe_zoffset;
+        #endif
+      #endif
```
Comment (thinkyhead) [inline Marlin/src/gcode/probe/M851.cpp]: Again, not valid C++ code and bad formatting.

File: Marlin/src/lcd/ultralcd.cpp [hunk 5]
```diff
@@ -1320,6 +1320,12 @@ void lcd_quick_feedback(const bool clear_buttons) {
           if (WITHIN(new_zoffset, Z_PROBE_OFFSET_RANGE_MIN, Z_PROBE_OFFSET_RANGE_MAX)) {
             thermalManager.babystep_axis(Z_AXIS, babystep_increment);
             zprobe_zoffset = new_zoffset;
+            #if ENABLED(DUAL_X_CARRIAGE) && ENABLED(INDEPENDENT_Z_OFFSETS)
+              if (active_extruder==0)
```
Comment (thinkyhead) [inline Marlin/src/lcd/ultralcd.cpp]: Operators must be preceded and followed by a single space.

File: Marlin/src/module/configuration_store.cpp [hunk 6]
```diff
@@ -494,13 +496,18 @@ void MarlinSettings::postprocess() {
...
-
     #if !HAS_BED_PROBE
       const float zprobe_zoffset = 0;
     #endif
-    EEPROM_WRITE(zprobe_zoffset);
-
+    #if ENABLED(INDEPENDENT_Z_OFFSETS) && ENABLED(DUAL_X_CARRIAGE)
+      _FIELD_TEST(x1_zprobe_zoffset);
+      EEPROM_WRITE(x1_zprobe_zoffset);
+      _FIELD_TEST(x2_zprobe_zoffset);
```
Comment (thinkyhead) [inline Marlin/src/module/configuration_store.cpp]: It's overkill to have a `_FIELD_TEST` for every field. A single instance would be fine here.

File: Marlin/src/module/configuration_store.cpp [hunk 7]
```diff
@@ -1180,13 +1187,22 @@ void MarlinSettings::postprocess() {
...
-
+      #if ENABLED(INDEPENDENT_Z_OFFSETS) && ENABLED(DUAL_X_CARRIAGE)
+        _FIELD_TEST(x1_zprobe_zoffset);
+        EEPROM_READ(x1_zprobe_zoffset);
+        _FIELD_TEST(x2_zprobe_zoffset);
+        EEPROM_READ(x2_zprobe_zoffset);
+        if(active_extruder==1)
+          zprobe_zoffset = x2_zprobe_zoffset;
+        else
+          zprobe_zoffset = x1_zprobe_zoffset;
```
Comment (thinkyhead) [inline Marlin/src/module/configuration_store.cpp]: For consistency, if you're using `active_extruder==1` in the other cases, it should also be used here. Comparison to zero also produces smaller code.

Keywords like `if` must be followed by a single space.

File: Marlin/src/module/configuration_store.cpp [hunk 8]
```diff
@@ -1906,7 +1922,10 @@ void MarlinSettings::reset(PORTARG_SOLO) {
   #if HAS_BED_PROBE
     zprobe_zoffset = Z_PROBE_OFFSET_FROM_EXTRUDER;
   #endif
-
+  #if ENABLED(INDEPENDENT_Z_OFFSETS) && ENABLED(DUAL_X_CARRIAGE)
+    x1_zprobe_zoffset = zprobe_zoffset;
+    x2_zprobe_zoffset = zprobe_zoffset;
```
Comment (thinkyhead) [inline Marlin/src/module/configuration_store.cpp]: Instead of using two variables, it would be better to have a single two-element array, `float zprobe_zoffset[2]`.

File: Marlin/src/module/motion.cpp [hunk 9]
```diff
@@ -1509,8 +1509,8 @@ void homeaxis(const AxisEnum axis) {
         }
         else {
           // In other modes, T0 can move from X_MIN_POS to X_MAX_POS
-          soft_endstop_min[axis] = base_min_pos(axis);
-          soft_endstop_max[axis] = base_max_pos(axis);
+          soft_endstop_min[axis] = X1_MIN_POS;
+          soft_endstop_max[axis] = X1_MAX_POS;
```
Comment (thinkyhead) [inline Marlin/src/module/motion.cpp]: Semantically it would be better to change these to `base_min_pos(X_AXIS)` and `base_max_pos(AXIS)`. The result will be the same. The `XYZ_CONSTS` macro might be good to redefine, however, replacing `const PROGMEM` with just `constexpr`. Try that and compare the resulting program size.

File: Marlin/src/module/probe.cpp [hunk 10]
```diff
@@ -51,6 +51,10 @@
 #endif
 
 float zprobe_zoffset; // Initialized by settings.load()
+#if ENABLED(DUAL_X_CARRIAGE) && ENABLED(INDEPENDENT_Z_OFFSETS)
+  float x1_zprobe_zoffset;
+  float x2_zprobe_zoffset;
```
Comment (thinkyhead) [inline Marlin/src/module/probe.cpp]: Should the single `zprobe_zoffset` still exist when there are also these? Perhaps not. Instead, it might be better to just make `zprobe_zoffset` into a two-element array and update all the code around that.

File: Marlin/src/module/configuration_store.cpp (hunk 8 above)
Comment (InsanityAutomation) [reply Marlin/src/module/configuration_store.cpp]: If we want to go this route, it may be better to make it an array all around and size it based on number of hotends. Then it could be independent for any multi tool machine, like a rotary system as well. That would no longer be idex specific and may need a bit more testing.

File: Marlin/src/module/probe.cpp (hunk 10 above)
Comment (InsanityAutomation) [reply Marlin/src/module/probe.cpp]: The single is what is live used to change the actual position. The other two swap with it behind the scenes as storage for that tool.

File: Marlin/src/module/tool_change.cpp [hunk 11]
```diff
@@ -495,14 +498,31 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
...
+              //if we dont have software endstops, use the configured limits instead
+              #if(active_extruder==0)
+                #if (!WITHIN(current_position[X_AXIS], X2_MIN_POS, X2_MAX_POS))
+                  no_move = true;
+                #endif
+              #elif (active_extruder==1)
+                #if (!WITHIN(current_position[X_AXIS], X1_MIN_POS, X1_MAX_POS))
+                  no_move = true;
+                #endif
+              #endif
```
Comment (thinkyhead) [inline Marlin/src/module/tool_change.cpp]: The above is not valid C++ code.

File: Marlin/src/module/tool_change.cpp [hunk 12]
```diff
@@ -511,8 +531,8 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
 
           set_destination_from_current();
 
-          const float xdiff = hotend_offset[X_AXIS][tmp_extruder] - hotend_offset[X_AXIS][active_extruder],
-                      ydiff = hotend_offset[Y_AXIS][tmp_extruder] - hotend_offset[Y_AXIS][active_extruder];
+          const float xdiff = hotend_offset[X_AXIS][tmp_extruder] - hotend_offset[X_AXIS][active_extruder];
+          const float ydiff = hotend_offset[Y_AXIS][tmp_extruder] - hotend_offset[Y_AXIS][active_extruder];
```
[the remark this answers is not in the document]
Comment (InsanityAutomation) [reply Marlin/src/module/tool_change.cpp]: I got a compile error for an undefined variable at one point without it. Arduino ide has been a bit unstable in recent releases though so it may be related to that.

File: Marlin/src/pins/pins_FORMBOT_TREX2+.h [hunk 13]
```diff
@@ -143,15 +143,13 @@
 #define HEATER_BED_PIN     58
 
 #define FAN_PIN             9
-//#define FAN1_PIN            4
-
+#if(DISABLED(FILAMENT_RUNOUT_SENSOR))
+  #define FAN1_PIN            4 //Dedicated laser port, can be repurposed as fil runout
+#endif
```
Comment (thinkyhead) [inline Marlin/src/pins/pins_FORMBOT_TREX2+.h]: This needs better formatting. Attention to detail inspires confidence.

File: Marlin/src/module/motion.cpp (hunk 9 above)
Comment (InsanityAutomation) [reply Marlin/src/module/motion.cpp]: Ill take a look at what sets those values. If those are not updated per x1 and x2, crash condition still exists.

File: Marlin/Configuration_adv.h [hunk 14]
```diff
@@ -373,6 +373,12 @@
       // without modifying the firmware (through the "M218 T1 X???" command).
       // Remember: you should set the second extruder x-offset to 0 in your slicer.
 
+      /* INDEPENDENT_Z_OFFSETS Allows setting a seperate z offset for X2. This allows compensation
+      * for tool heads that are at different heights. Please note this will be ignored in duplicate mode
+      * and the default z probe offset will be used.
+      */
+  //#define INDEPENDENT_Z_OFFSETS
```
Comment (thinkyhead) [inline Marlin/Configuration_adv.h]: - Please follow the established standards for comment formatting and indentation.
- The word "separate" is misspelled.
- "z" should be capitalized.

Admittedly the comments for `DUAL_X_CARRIAGE` are pretty terrible. I will clean up the others.

File: Marlin/src/module/configuration_store.cpp [hunk 15]
```diff
@@ -154,7 +154,9 @@ typedef struct SettingsDataStruct {
   // HAS_BED_PROBE
   //
   float zprobe_zoffset;                                 // M851 Z
-
+  #if ENABLED(INDEPENDENT_Z_OFFSETS) && ENABLED(DUAL_X_CARRIAGE)
+    float x2_zprobe_zoffset;
+  #endif
```
Comment (thinkyhead) [inline Marlin/src/module/configuration_store.cpp]: If there is to be either `zprobe_zoffset` or `x1/x2_zprobe_zoffset` but not both, that should also be reflected in the structure here.

This `#endif` should be followed by a blank line.

File: Marlin/src/module/configuration_store.cpp (hunk 8 above)
Comment (thinkyhead) [reply Marlin/src/module/configuration_store.cpp]: Makes sense.

I'm still not very clear how two probes is going to work. You have such a setup?

File: Marlin/src/module/tool_change.cpp (hunk 11 above)
Comment (InsanityAutomation) [reply Marlin/src/module/tool_change.cpp]: Updated

File: Marlin/src/pins/pins_FORMBOT_TREX2+.h (hunk 13 above)
Comment (InsanityAutomation) [reply Marlin/src/pins/pins_FORMBOT_TREX2+.h]: Updated

File: Marlin/src/module/motion.cpp (hunk 9 above)
Comment (thinkyhead) [reply Marlin/src/module/motion.cpp]: The `base_...` macros are just aliases to get items from a `const` array. If there's anything weird going on, then to my mind it tends to implicate the use of `PROGMEM` in the definition of `XYZ_CONSTS`.

File: Marlin/src/module/configuration_store.cpp (hunk 15 above)
Comment (InsanityAutomation) [reply Marlin/src/module/configuration_store.cpp]: Updated, originally just looked at it as ensuring size matched, and the first 2 are used interchangeably in the save / restore scenario.

File: Marlin/src/module/probe.cpp (hunk 10 above)
Comment (thinkyhead) [reply Marlin/src/module/probe.cpp]: Menus are totally ok with using something like `zprobe_zoffset[active_extruder]` as the thing being live-edited.

File: Marlin/src/module/configuration_store.cpp [hunk 16]
```diff
@@ -153,7 +153,14 @@ typedef struct SettingsDataStruct {
...
   //
-  float zprobe_zoffset;                                 // M851 Z
+                                   // M851 Z
+  if (ENABLED(INDEPENDENT_Z_OFFSETS) && ENABLED(DUAL_X_CARRIAGE)) {
+    float x1_zprobe_zoffset;
+    float x2_zprobe_zoffset;
+  }
+  else {
+    float zprobe_zoffset;
+  }
```
Comment (thinkyhead) [inline Marlin/src/module/configuration_store.cpp]: This is not valid C++.

File: Marlin/src/module/planner.cpp [hunk 17]
```diff
@@ -1414,7 +1414,7 @@ void Planner::check_axes_activity() {
   }
 
   void Planner::unapply_leveling(float raw[XYZ]) {
-
+    planner.synchronize();
```
Comment (thinkyhead) [inline Marlin/src/module/planner.cpp]: No need to use `planner.` prefix within a `Planner::` method.

File: Marlin/src/module/configuration_store.cpp (hunk 8 above)
Comment (InsanityAutomation) [reply Marlin/src/module/configuration_store.cpp]: Just one probe, but the nozzles are physically at different heights, therefore each nozzle needs a different offset from the same probe.

File: Marlin/src/module/motion.cpp (hunk 9 above)
Comment (InsanityAutomation) [reply Marlin/src/module/motion.cpp]: The values I placed in are config file values from the dual x carriage segment that limit the travel of the specific tool head, so they may not match. In this case, the base_max_pos for x axis is 442, but the last 40mm is the park position of X2 so x1_max_pos is 400. This prevents it from trying to go through X2 while parking, or allowing you to jog x1 into x2.

File: Marlin/src/module/tool_change.cpp [hunk 18]
```diff
@@ -56,6 +56,9 @@
 #if HAS_FANMUX
   #include "../feature/fanmux.h"
 #endif
+#if ENABLED(NEWPANEL) || ENABLED(ULTIPANEL)
```
[the remark this answers is not in the document]
Comment (thinkyhead) [reply Marlin/src/module/tool_change.cpp]: In that case, just `ULTIPANEL`. That define essentially indicates that programmable menus exist.

File: Marlin/src/module/probe.cpp (hunk 10 above)
Comment (InsanityAutomation) [reply Marlin/src/module/probe.cpp]: But if tool change is called while the screen is up, you wont see the value for the new extruder until the first change is made. I wanted to drop out of the menu so that you dont see a stale value.

File: Marlin/src/module/tool_change.cpp [hunk 19]
```diff
@@ -584,6 +604,15 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
...
 
       planner.synchronize();
+      SERIAL_ECHOLNPAIR(MSG_ACTIVE_EXTRUDER, (int)active_extruder);
+      #if (ENABLED(DUAL_X_CARRIAGE) && ENABLED(INDEPENDENT_Z_OFFSETS))
+        if (active_extruder==1) {
+          zprobe_zoffset = x2_zprobe_zoffset;
+        }
+        else {
+            zprobe_zoffset = x1_zprobe_zoffset;
+        }
```
[the remark this answers is not in the document]
Comment (InsanityAutomation) [reply Marlin/src/module/tool_change.cpp]: Personally dont care for this style but its used alot so ill change it here.

File: Marlin/src/module/probe.cpp (hunk 10 above)
Comment (thinkyhead) [reply Marlin/src/module/probe.cpp]: If there are going to be two independent offsets, it will be best to have two menu items — one for X1 and one for X2.

File: Marlin/src/module/planner.cpp (hunk 17 above)
Comment (InsanityAutomation) [reply Marlin/src/module/planner.cpp]: Updated

File: Marlin/src/module/tool_change.cpp [hunk 20]
```diff
@@ -495,14 +499,28 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
 
           #if HAS_SOFTWARE_ENDSTOPS
             // Update the X software endstops early
+            const float old_extruder = active_extruder;
             active_extruder = tmp_extruder;
             update_software_endstops(X_AXIS);
-            active_extruder = !tmp_extruder;
+            active_extruder = old_extruder;
```
Comment (thinkyhead) [inline Marlin/src/module/tool_change.cpp]: Fine, but an unnecessary change. We know that the tool is definitely being changed at this point, and we know that 0 and 1 are the only valid values for `tmp_extruder`.

File: Marlin/src/module/configuration_store.cpp (hunk 8 above)
Comment (thinkyhead) [reply Marlin/src/module/configuration_store.cpp]: The Z probe Z offset is only used during probing. And we already know the second nozzle's Z offset from `hotend_offset[Z_AXIS][1]`. If there is no second probe attached to the second nozzle, then there's nothing useful being added by a second Z probe Z offset.

File: Marlin/src/module/tool_change.cpp (hunk 20 above)
Comment (InsanityAutomation) [reply Marlin/src/module/tool_change.cpp]: For now, but i figured this may be more future-proof if anything adds more down the road.

**Comment by thinkyhead** [comment]
Rebased and squashed. To get the updated branch, you must do the following in the Git console:
```
git fetch origin
git checkout IndependentZOffsetV2
git reset --hard origin/IndependentZOffsetV2
```

File: Marlin/src/module/motion.cpp (hunk 9 above)
Comment (thinkyhead) [reply Marlin/src/module/motion.cpp]: The max position for X1 should already be configured so that it can only go as far as being next to X2. But, since the X1_MIN_POS and X1_MAX_POS can be overridden, your change here is fine.

**Comment by Roxy-3D** [comment]
Guys...   Please don't make me go to the security logs to see what is being deleted.   Let's just have a respectful conversation about the technical merits of the Pull Request.

File: Marlin/src/module/tool_change.cpp (hunk 20 above)
Comment (thinkyhead) [reply Marlin/src/module/tool_change.cpp]: Compile size takes precedence. We can update it when it _needs_ to be updated.

**Comment by thinkyhead** [comment]
I'm deleting comments on dead commits that no longer apply. There's nothing weird going on.

**Comment by Roxy-3D** [comment]
OK.  I apologize.   It looked like some crazy stuff was happening.

**Comment by InsanityAutomation** [comment]
@Roxy-3D also some behind the scenes discussion on the discord. Nothing too crazy.

**Comment by thinkyhead** [comment]
The only crazy thing is that deleting old commit comments didn't help make the page any shorter or neater at all, so it was a waste of time. 😞

**Comment by thinkyhead** [comment]
True. All the best name-calling is happening on the Discord.

**Comment by InsanityAutomation** [comment]
We need to make the page long enough a printed copy rivals the Guinness book of world records. That in an of itself will become a record, make the book longer, the thread will need to be longer, therefore infinite loop of updates.

**Comment by Roxy-3D** [comment]
No...   short and concise is better.   We don't need an entry in the Guinness book of records.

**Comment by thinkyhead** [comment]
The longest thread ever award already goes to the layer shift bug thread. I think it's still going even though the main issue was long ago fixed.

**Comment by Roxy-3D** [comment]
> The longest thread ever award already goes to the layer shift bug thread. I think it's still going.

Actually....   Isn't #7076 or its pre-cursor thread asking for recommendations on how to get to a 32-bit platform longer?   Both of those threads are monstrous.

And...  if the pre-cursor thread had not been stopped and #7076 started....  All of that stuff would be in one bigger thread.

**Comment by InsanityAutomation** [comment]
@Roxy-3D Weve decided to use the hotend offsets moving in opposite directions to remove the need for eeprom changes, and make sure its always active immediately. Ill make the changes tomorrow, 1a here lol

And see, we made it longer again!

**Comment by bruce356** [comment]
@InsanityAutomation hi, has Z-Height problem with tool changes back and forth also been fixed?
Thanks - bruce

**Comment by InsanityAutomation** [comment]
There are a couple of things between the last few commits we've been working on that may help that. I've gotta finish up the most recent pull request then we can have you try again. If that doesn't work, I'll give you a branch with some serial debug data added, and get a log from you.

**Comment by InsanityAutomation** [comment]
@Roxy-3D care to give this method a try? I'll need to put in a more permanent fix for the hotend offsets being applied during tool change.

**Comment by Roxy-3D** [comment]
I'm trying to get the current code base running with IDEX enabled.   There is some problem with the endstops right now.  I have the reporting logic working again.   But the homing fails because it never sees the endstop get hit.

**Comment by InsanityAutomation** [comment]
Ouch, I'm still working from a week old branch. If it's not sorted when I sit back down to clean up what I've got here and pull the eeprom changes, I can take a look at that next.

**Comment by Roxy-3D** [comment]
The reporting logic for M119 was 5 days ago.   I believe I have that fixed.   If I was more Git capable...  I could probably look for other endstop logic that got changed 5 days ago.    That would be quick and easy if I knew my way around Git.

But I don't....   Do you know Git very well?  If so...   It might just take you a second to find where that got corrupted.

**Comment by InsanityAutomation** [comment]
Mostly I'd just use the history tab in git desktop, or look at changes from my working branch till now.

**Comment by Roxy-3D** [comment]
Yeah...  And I'm starting to waffle.   I backed up to a branch I had been using that is a week old...  And it is dying on the G28 home also the same way.   It is almost like something went wrong with my hardware.

**Comment by InsanityAutomation** [comment]
You had a trex2+ right? You can try my stable branch for tinymachines to rule that out - https://github.com/InsanityAutomation/Marlin/tree/TM_Trex2%2B_2.0.x?files=1

**Comment by Roxy-3D** [comment]
I'll try that now...

Aaaarrrrgghhhhh...   You probably compile with PlatformIO, right?   Because the duplicate configuration.h files messes everything up with Arduino.

**Comment by InsanityAutomation** [comment]
@Roxy-3D actually, Arduino ide, not platformio. Did you ever get something working?

**Comment by Roxy-3D** [comment]
> actually, Arduino ide, not platformio. Did you ever get something working?

Yes...   It turns out I had a blown step stick.   It took some effort to figure that out because the bad step stick was sucking a lot of the power out of the other step sticks.   Once I started swapping step sticks, things started to make sense.   But then I still had to get the new step stick with a glued on heat sink.  And then get the current set at max without it doing thermal shut down.   Yadi... yadi... yadi...  Nothing ever is easy.   You know the drill.

Any way...  I got my machine working again.  Except...  now I'm back to where I started.   And I'm trying to upgrade it so it is a T-Rex 3.   And I don't have all the mounting brackets I need so I'm having to make stuff.   uggggh!

**Comment by InsanityAutomation** [comment]
I know the drill! Stock stepper cooling is rather weak. Did you use a serial connection and the otp bits to set the y stepper to spreadcyle? If not be prepared for some issues with skipped steps and thermal shutdown under load.

**Comment by Roxy-3D** [comment]
No...  I kept lowering the reference voltage until it didn't do a thermal shut down.  And then a lowered the reference voltage a little bit more.    I guess we will see how that works out.   But did have the extruder moving back and forth for 20 minutes and no steps were lost.  So maybe I'll be OK.

**Comment by InsanityAutomation** [comment]
X axis was fine on stealthchop. Y is where I had a problem on long jobs. Couldn't get past a few hours with anything unless I slowed it way down. Lower it enough to avoid thermal shutdown, I got layer shifts.

**Comment by Roxy-3D** [comment]
The thing is...   Gear ratio reducing mechanisms that bolt on to any stepper motor are available.   I don't understand why they aren't in use every where.   It shouldn't be necessary to have the current cranked up so high so the stepper motors don't lose steps.

**Comment by InsanityAutomation** [comment]
I agree, I think many don't want to sacrifice speed though. In some cases it probably wouldn't make a difference....

**Comment by InsanityAutomation** [comment]
@Roxy-3D should be pretty much together now. Ill take another poke at the dual x tool change offset application later on to clean that up, but I may make that a separate pull req. This one already has more independent things than id like in a single commit lol

**Comment by Roxy-3D** [comment]
It will be several days (if everything goes smoothly!!!!) before I can test anything.   But this has already been scrubbed by @thinkyhead so if it is helpful to merge it, I will do that.

**Comment by thinkyhead** [comment]
This PR is far from ready. I’ll post a follow-up soon.

**Comment by silentninja1** [comment]
@thinkyhead yeah, id say changed to more in line with what we were talking about, and open for another set of eyes to try it functionally. Now that its not idex specific, id like to know how it performs on a switching extruder or something as well. I dont have one of those.

**Comment by Roxy-3D** [comment]
> This PR is far from ready. I’ll post a follow-up soon.

Sorry!  I looked at it and didn't see any problems....

**Comment by silentninja1** [comment]
@Roxy-3D and I quote - ThinkyheadYesterday at 8:39 PM
Git and the alpha designation on 2.0.x make all mistakes tolerable.

:) I wont claim to know better than either of you when something is up to snuff to commit. But ill take whatever feedback I get and make it as close as I can!

**Comment by thinkyhead** [comment]
@silentninja1 — Certain mistakes are tolerable. Take a look at #11916 to see the amount of followup that was required to get this PR into proper shape. A good example of what stretches the limits of our patience and tolerance.
